src/configuration_server.erl

```diff
@@ -0,0 +1,155 @@
+init([]) ->
+    case config_file_name() of
+        {ok, ConfigFilePath} -> {ok, #configuration{config_file_path = ConfigFilePath}};
+        {error, Reason} -> {stop, Reason}
+    end.
+
+%% @private
+%% @doc Handling call messages
+-spec handle_call(
+    Request :: term(),
+    From :: {pid(), Tag :: term()},
+    State :: #configuration{}
+) ->
+    {reply, Reply :: term(), NewState :: #configuration{}} |
+    {reply, Reply :: term(), NewState :: #configuration{}, timeout() | hibernate} |
+    {noreply, NewState :: #configuration{}} |
+    {noreply, NewState :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term(), Reply :: term(), NewState :: #configuration{}} |
+    {stop, Reason :: term(), NewState :: #configuration{}}.
+handle_call(config_file, _From, State = #configuration{}) ->
+    {reply, State#configuration.config_file_path, State};
+handle_call(default_directories, _From, State = #configuration{}) ->
+    {reply, {ok, default_directories()}, State};
+handle_call(_Request, _From, State = #configuration{}) ->
+    {reply, ok, State}.
+
+%% @private
+%% @doc Handling cast messages
+-spec handle_cast(Request :: term(), State :: #configuration{}) ->
+    {noreply, NewState :: #configuration{}} |
+    {noreply, NewState :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #configuration{}}.
+handle_cast(_Request, State = #configuration{}) ->
+    {noreply, State}.
+
+%% @private
+%% @doc Handling all non call/cast messages
+-spec handle_info(Info :: timeout() | term(), State :: #configuration{}) ->
+    {noreply, NewState :: #configuration{}} |
+    {noreply, NewState :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #configuration{}}.
+handle_info(_Info, State = #configuration{}) ->
+    {noreply, State}.
+
+%% @private
+%% @doc This function is called by a gen_server when it is about to
+%% terminate. It should be the opposite of Module:init/1 and do any
+%% necessary cleaning up. When it returns, the gen_server terminates
+%% with Reason. The return value is ignored.
+-spec terminate(
+    Reason :: (normal | shutdown | {shutdown, term()} | term()),
+    State :: #configuration{}
+) ->
+    term().
+terminate(_Reason, _State = #configuration{}) ->
+    ok.
+
+%% @private
+%% @doc Convert process state when code is changed
+-spec code_change(
+    OldVsn :: term() | {down, term()},
+    State :: #configuration{},
+    Extra :: term()
+) ->
+    {ok, NewState :: #configuration{}} | {error, Reason :: term()}.
+code_change(_OldVsn, State = #configuration{}, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+%% @doc This function returns the predefined configuration directories.
+-spec default_directories() ->
+    [file:filename_all()].
+default_directories() ->
+    ?CONFIG_PATHS.
```

Comment:
To the best of my knowledge, there are nor specs for macros, since macros are C-like test replacements (nothing fancy, like Lisp macros). The implementation might change, but this function should return a list of filenames/directories.

Reply:
Also kinda bugs me that there's another pair of functions named with a noun.